src/pages/ReimbursementAccount/BeneficialOwnersStep.js

```diff
@@ -53,16 +55,31 @@ class BeneficialOwnersStep extends React.Component {
         };
     }
 
+    /**
+     * @returns {Object}
+     */
+    getErrors() {
+        return lodashGet(this.props, ['reimbursementAccount', 'errors'], {});
+    }
+
+    /**
+     * @param {Integer} ownerIndex
+     * @returns {Object}
+     */
+    getBeneficialOwnerErrors(ownerIndex) {
+        return lodashGet(this.getErrors(), `beneficialOwnersErrors[${ownerIndex}]`, {});
```

Comment:
Do we actually need to validate every field in a beneficial owners? I don't think we ever really confirmed that this was necessary or not and might be good to clarify.

Reply:
Oh, I didn't confirm, I assumed we were validating them because the previous code was validating them:

https://github.com/Expensify/App/blob/8db0dd072c0d5193bf045bdc49187b3d2c551c26/src/pages/ReimbursementAccount/BeneficialOwnersStep.js#L60-L67

Reply:
Yeah, that's a good assumption and seems right, but I'm not sure what I'd expect - should we show inline fields for every field on a per owner basis? Or should we just tell them something is wrong with the "beneficial owners" and treat each owner like a form field too? 

It's not a "form field" but it is a "form inside a form". Anyway, I'm trying to think about how we would achieve this if we had something like a reusable form component (that we are planning to do in a design doc). And with the code we have so far here... it seems very specific to have things like `clearErrorAndSetBeneficialOwnerValue()`.

Gonna tag in @luacmartins to see if he has any thoughts on how we might solve this.

Reply:
hmm, given that we are storing errors in Onyx, this way made sense to me and we are still able to reuse the `IdentityForm` component and validations on it from `ValidationUtils.js`. I also understand that it got a bit complex here because we handle storing and clearing errors outside of each `IdentityForm`.

An alternative approach could be to give the responsibility of storing and clearing errors to the `IdentityForm` component. The errors would be stored in a local state, not in Onyx.

Reply:
>I also understand that it got a bit complex here because we handle storing and clearing errors outside of each IdentityForm.

Makes sense. This is all in flux and I'm not sure what would be best. But can think of a few ways to handle this:

1. Do what you have here and pass in errors to an "inner form" letting the parent control set which errors show with custom logic.

2. Come up with a general convention that a form should avoid having a "form inside a form" and if we must have something like that then we should be able to handle dynamic form fields. And instead of an arrays of owners we'd have something like `formData.beneficialOwner_1.someField` etc.

3. We will allow forms inside forms, but have validation for the inner form happen independently from the outer form. This way we do not need to set errors for dynamic fields but maybe when we submit the inner form will make the main validation for the outer form fail.

I'm honestly not too sure which design would be better or if there's another one we could do instead.

If we look at how this is handled in Web-Secure it seems like the approach leans more towards `3` where the IdentityForm component has [validation on each field](https://github.com/Expensify/Web-Secure/blob/c630d65f1c31ca6b87de9d0299a7a6bf37e6767d/site/app/dialogs/reimbursementAccount/IdentityForm.jsx#L131-L136) and then we check if there are any errors [here and block submit](https://github.com/Expensify/Web-Secure/blob/c630d65f1c31ca6b87de9d0299a7a6bf37e6767d/site/app/dialogs/reimbursementAccount/beneficialOwnersStep.jsx#L43-L71).

Reply:
I'd expect inline errors for each beneficial owner here vs one generic error for the whole `IndentityForm`. I _think_ that the current way of handling errors with Onyx and managing it outside of `IdentityForm` seems less complex/confusing than moving it inside `IdentityForm` and passing references around. I guess instead of a form inside a form, I'm thinking of it as a dynamic form, where we can add/remove fields and all data/errors are managed in the main form.

 I would imagine that the new form design could benefit from using a similar structure to the draft data we have, e.g. `beneficialOwners: [{...owner1}, {...owner2}]` and each input in `IdentityForm` manipulating those data and error values directly.

Reply:
Just read @marcaaron latest comment and I think my proposed solution above aligns with 2, with the caveat of the form structure. Having `formData.beneficialOwner_1.someField` would work well too.

Reply:
Hmm, I could implement number 2 quickly in a separate branch to see how it looks, what do you think?

Reply:
I think I was miss-reading number 2.

Is the difference between (1) and (2) the following?

### (1) current approach, keep fields related to a beneficial owner in a separate object, these objects are put together in an array
```
{
   ...otherFieldData, // Not related to beneficial owners
   beneficialOwners: [
       { ...owner1Fields },
       { ...owner2Fields },
   ]
}
```

### (2) keep fields related to a beneficial owner in a separate object, these objects are associated to a key directly in the main form object
```
{
   ...otherFieldData, // Not related to beneficial owners
   beneficialOwners_1: { ...owner1Fields },
   beneficialOwners_2: { ...owner2Fields },
}
```

I feel like there is not much difference between them, custom code will be required anyway.

There is also another approach, but can be ugly. If we don't allow "nested" forms, we could force the beneficial owners' fields to be laid out in a flat object...

```
{
   ...otherFieldData,
   beneficialOwners_1_address,
   beneficialOwners_1_city,
   beneficialOwners_1_<etc>,
   beneficialOwners_2_address,
   beneficialOwners_2_city,
   beneficialOwners_2_<etc>,
}
```
This could make it easier to reuse our functions for cleaning up errors and setting values I think. But I also find it dirty to be creating this kind of keys and parsing them.

Reply:
Alright sounds like we are leaning towards:

- **Yes** we should show inline errors (I think I agree with this but was mainly stuck on the best way to implement)
- It would be better to treat the `IdentityForm` as a something like a bunch of form inputs together instead of another form and pass the errors in like you are already doing @aldo-expensify (I agree this feels less confusing than a form in a form style).

The only thing I'm not too sure about is whether we will need `clearErrorAndSetBeneficialOwnerValue()`. I think maybe using a new key for each owner could help here since I'd imagine that whatever "higher design" we rely on in the future to collect form errors would be a simple data structure. If we do that then we can just use the same method here:

https://github.com/Expensify/App/blob/2036c3fb2631f2de50bf9061f7c38f9146aaf04f/src/libs/ReimbursementAccountUtils.js#L30-L41

and apply it like we are doing here:

https://github.com/Expensify/App/blob/2036c3fb2631f2de50bf9061f7c38f9146aaf04f/src/pages/ReimbursementAccount/CompanyStep.js#L101-L118

Instead of having to create this method that we need to accommodate the array structure.

What do we think?

Reply:
>There is also another approach, but can be ugly. If we don't allow "nested" forms, we could force the beneficial owners' fields to be laid out in a flat object...

Yeah this is what I had in mind sorry for missing that the same problem would exist whether we have an array or nested object. A simple flat structure was what I had in my head. Or maybe there's a way to work with a more complex structure, but if we allow it then it should be more generic somehow? 

I think more than anything... we are going to re-design everything about this so maybe it's not a very important problem to solve now and we should take the path that requires the least amount of code.

Reply:
Oh I had a random idea. Unsure if it's perfect but if we do nothing and allow any data structure so keep something like this...

```js
errors: {
    beneficialOwners: [
        {field1: true, field2: true, etc},
    ],
},
```

But maybe our methods for `setValue()` and `clearError()` become able to work with nested values similar to `lodashGet()`

So we could do 

```js
clearErrorAndSetValue(['beneficialOwners', index, 'phone'], value);
```

Reply:
> I think more than anything... we are going to re-design everything about this so maybe it's not a very important problem to solve now and we should take the path that requires the least amount of code.

+1

> But maybe our methods for setValue() and clearError() become able to work with nested values similar to lodashGet()

I like this! I think it's a great solution to handle complex data structures in the forms and would also give us the most flexibility to design the component!

Reply:
>I think more than anything... we are going to re-design everything about this so maybe it's not a very important problem to solve now and we should take the path that requires the least amount of code.

Yep, thought about that too. Maybe we should spend to much time designing this here because it may be incompatible with we may choose in the future.

Reply:
>But maybe our methods for setValue() and clearError() become able to work with nested values similar to lodashGet()



I'll give this a try

Reply:
Ok.. committed a first version of a function that works like [lodash-set](https://lodash.com/docs/4.17.15#set) (lodash-set is for setting a value given a path, similar to the lodash-get) and handles immutability.

I used it in the `BeneficialOwnersStep.js` and indeed the very specific code to set reset the error was simplified greatly.

If we like this approach, I could use it for updating also the values of the form (I just used it now for clearing errors).

See commit https://github.com/Expensify/App/pull/5259/commits/c658ad10989b620d850e15382a029e9323f903a0

Reply:
>Sadly, lodash-set doesn't handle immutability

Can you explain this problem some more? I think we can just clone the object and then use lodash-set on it? Won't that be fine in most cases? What are we worried will happen?

Reply:
Oh, cloning the object is a simpler alternative to handle mutability, I just didn't do it like that for performance reasons, but maybe that doesn't matter here because objects are small.

Reply:
Nice, I think this is resolve now. 🎉